### PR TITLE
Add limits

### PR DESCRIPTION
**Pull request by Chepelau**

### CODE REVIEW

File: contracts/constants.cairo [hunk 1]
```diff
@@ -22,7 +22,7 @@ const ACCOUNT_BALANCE_UPPER_BOUND = 2 ** 64 * Math64x61.FRACT_PART;
 // The minimum and maximum volatility
 const VOLATILITY_LOWER_BOUND = 1;
 const VOLATILITY_UPPER_BOUND = 2 ** 64 * Math64x61.FRACT_PART;
-
+const MAX_OPTION_SIZE_PERCENT_OF_VOLADJSPD = 20;
```
Comment (MarekHauzr) [inline contracts/constants.cairo:25]: put commnet above describing what this is

Comment (Chepelau) [reply contracts/constants.cairo:25]: also, wouldn't it be better to make it adjustable? so we can effectively turn it off in tests??? otherwise we have to rewrite bunch of them

File: contracts/liquidity_pool.cairo [hunk 2]
```diff
@@ -413,6 +413,11 @@ func deposit_liquidity{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_che
         set_lpool_balance(lptoken_address, new_pb);
     }
 
+    with_attr error_message("lpool balance exceeds set maximum"){
```
Comment (MarekHauzr) [inline contracts/liquidity_pool.cairo:416]: please add a comment why this is needed (what it is used for)... the inside... the max_balance cap...

File: contracts/state.cairo [hunk 3]
```diff
@@ -360,6 +365,20 @@ func get_pool_volatility_auto{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, ra
 }
 
 
+@view
+func get_max_lpool_balance{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    pooled_token_addr: Address) -> (max_balance: Uint256) {
+        let(maxbal) = max_lpool_balance.read(pooled_token_addr);
+        return (maxbal, );
```
Comment (MarekHauzr) [inline contracts/state.cairo:372]: add assert that the maxbal is positive value

File: contracts/state.cairo [hunk 4]
```diff
@@ -360,6 +365,20 @@ func get_pool_volatility_auto{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, ra
...
+}
+
+@external
+func set_max_lpool_balance{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    pooled_token_addr: Address, max_lpool_bal: Uint256) {
+        max_lpool_balance.write(pooled_token_addr, max_lpool_bal);
```
Comment (MarekHauzr) [inline contracts/state.cairo:378]: add assert that the max_lpool_bal is positive

File: contracts/state.cairo [hunk 5]
```diff
@@ -392,7 +411,7 @@ func set_pool_volatility{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_c
     return ();
 }
 
-
+@external
```
Comment (MarekHauzr) [inline contracts/state.cairo:414]: nope... no @external here

File: contracts/options.cairo [hunk 6]
```diff
@@ -165,6 +166,19 @@ func mint_option_token{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_che
...
+
+        with_attr error_message("trade exceeds max liqpool deposit size"){
+            let (user_address) = get_caller_address();
+            let (currency_address) = get_underlying_token_address(lptoken_address);
+            let (optsize) = IOptionToken.balanceOf(contract_address=option_token_address, account=user_address);
+            assert optsize.high = 0;
+            let (adjspd) = get_pool_volatility_adjustment_speed(lptoken_address=lptoken_address);
+            let adjspd_ = adjspd * 100;
+            assert_nn(adjspd_);
+            let max_optsize_m64x61 = adjspd_ /  MAX_OPTION_SIZE_PERCENT_OF_VOLADJSPD;
```
Comment (MarekHauzr) [inline contracts/options.cairo:178]: shouldn't it be `*` instead of `/` since the MAX_OPTION_SIZE_PERCENT_OF_VOLADJSPD in %? or even better

`adjspd_ *  MAX_OPTION_SIZE_PERCENT_OF_VOLADJSPD / 100`

File: contracts/constants.cairo (hunk 1 above)
Comment (MarekHauzr) [reply contracts/constants.cairo:25]: hmmm... I guess you are right... could be done similarly like `volatility_adjustment_speed: Math64x61_` in `add_lptoken`

File: contracts/state.cairo [hunk 7]
```diff
@@ -360,6 +365,20 @@ func get_pool_volatility_auto{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, ra
...
 
 
+@view
+func get_max_lpool_balance{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    pooled_token_addr: Address) -> (max_balance: Uint256) {
+        let(maxbal) = max_lpool_balance.read(pooled_token_addr);
+        return (maxbal, );
+}
+
+@external
```
Comment (MarekHauzr) [inline contracts/state.cairo:375]: should not be external without `Proxy.assert_only_admin();`

also, create one `set_max_lpool_balance` that is not external and is used in the add_lptoken... and another one `set_max_lpool_balance_external` that is used for updating existing pools...

similarly to how `set_pool_volatility_adjustment_speed_external` is done

Comment (Chepelau) [reply contracts/state.cairo:375]: also i don't see the function in the interface

Comment (MarekHauzr) [reply contracts/state.cairo:375]: interface may be behind

File: contracts/options.cairo [hunk 8]
```diff
@@ -185,6 +186,25 @@ func mint_option_token{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_che
                 strike_price=strike_price,
             );
         }
+
+        with_attr error_message("trade exceeds max liqpool deposit size"){
+            let (user_address) = get_caller_address();
+            let (optsize) = IOptionToken.balanceOf(contract_address=option_token_address, account=user_address);
```
Comment (MarekHauzr) [inline contracts/options.cairo:192]: I wouldn't block the amount of option tokens the user has... just the amount the user mints at one go... cause that is the impact on the AMM

File: contracts/options.cairo [hunk 9]
```diff
@@ -185,6 +186,25 @@ func mint_option_token{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_che
...
+
+            let max_opt_perc_math = Math64x61.fromFelt(MAX_OPTION_SIZE_PERCENT_OF_VOLADJSPD);
+            let hundred = Math64x61.fromFelt(100);
+            let ratio = Math64x61.div(max_opt_perc_math, hundred);
+
+            let max_optsize_m64x61 = Math64x61.mul(ratio, adjspd);
+
+            let max_optsize = toInt_balance(max_optsize_m64x61, option_token_address);
+            assert_le(optsize.low, max_optsize);
+        }
```
Comment (MarekHauzr) [inline contracts/options.cairo:207]: also this same thing should be in the burn_option token... ie burn is closing of the position and the impact on the amm is the same as opening...

File: contracts/state.cairo [hunk 10]
```diff
@@ -70,6 +70,34 @@ func option_token_address(
...
+}
+
+
+// DEPRECATED
+@storage_var
+func option_position(
+    lptoken_address: Address, option_side: OptionSide, maturity: Int, strike_price: Math64x61_
+) -> (res: Math64x61_) {
+}
+
```
Comment (MarekHauzr) [inline contracts/state.cairo:100]: why is this added???

File: contracts/state.cairo [hunk 11]
```diff
@@ -70,6 +70,34 @@ func option_token_address(
 ) -> (res: Address) {
 }
 
+@storage_var
+func max_lpool_balance(pooled_token_addr: Address) -> (res: Uint256) {
+}
+
+
+@external
+func migrate_option_position{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
```
Comment (MarekHauzr) [inline contracts/state.cairo:79]: why is htis added? wasn't it already migrated?

File: contracts/state.cairo (hunk 5 above)
Comment (MarekHauzr) [reply contracts/state.cairo:462]: still external... the `set_pool_volatility_separate` CANNOT BE EXTERNAL

File: contracts/state.cairo (hunk 7 above)
Comment (MarekHauzr) [reply contracts/state.cairo:398]: please take a look into this

File: contracts/state.cairo (hunk 10 above)
Comment (Chepelau) [reply contracts/state.cairo:100]: no idea @tensojka

File: contracts/options.cairo [hunk 12]
```diff
@@ -162,6 +162,9 @@ func mint_option_token{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_che
             assert contract_option_side = option_side;
         }
 
+        let (user_address) = get_caller_address();
+        let (optsize_before) = IOptionToken.balanceOf(contract_address=option_token_address, account=user_address);
```
Comment (MarekHauzr) [inline contracts/options.cairo:166]: I think it should be checked against `option_size` and not the balance... the user can go around by having multiple addresses... so neither is fool proof, but I prefer the option_size

if you disagree, just comment that you disagree

File: contracts/options.cairo [hunk 13]
```diff
@@ -185,6 +188,27 @@ func mint_option_token{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_che
                 strike_price=strike_price,
             );
         }
+
+        with_attr error_message("trade exceeds max liqpool deposit size"){
+            let (optsize_after) = IOptionToken.balanceOf(contract_address=option_token_address, account=user_address);
+            assert optsize_after.high = 0;
+            
+            let (optsize_diff: Uint256) = uint256_sub(optsize_after, optsize_before);
```
Comment (MarekHauzr) [inline contracts/options.cairo:196]: omg... why are you calling balanceOf twice? you could have just used option_size... please use option_size instead of optsize_diff and remove the balanceOf calls

File: contracts/options.cairo [hunk 14]
```diff
@@ -185,6 +188,27 @@ func mint_option_token{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_che
                 strike_price=strike_price,
             );
         }
+
+        with_attr error_message("trade exceeds max liqpool deposit size"){
```
Comment (MarekHauzr) [inline contracts/options.cairo:192]: `trade exceeds max liqpool deposit size` shouldn't be something like `trade exceeds max option size`... or something like this?

File: contracts/options.cairo [hunk 15]
```diff
@@ -587,6 +614,27 @@ func burn_option_token{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_che
             );
         }
     }
+
+    with_attr error_message("trade exceeds max liqpool deposit size"){
```
Comment (MarekHauzr) [inline contracts/options.cairo:618]: ditto

File: contracts/options.cairo [hunk 16]
```diff
@@ -587,6 +614,27 @@ func burn_option_token{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_che
...
+        let (optsize_after) = IOptionToken.balanceOf(contract_address=option_token_address, account=user_address);
+        assert optsize_after.high = 0;
+        
+        let (optsize_diff) = uint256_sub(optsize_before, optsize_after);
```
Comment (MarekHauzr) [inline contracts/options.cairo:622]: ditto

File: contracts/state.cairo [hunk 17]
```diff
@@ -360,6 +368,51 @@ func get_pool_volatility_auto{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, ra
...
+    
+        with_attr error_message("max lpool balance can't be negative") {
+            assert_uint256_le(Uint256(0, 0), max_lpool_bal);
+        }
+        max_lpool_balance.write(pooled_token_addr, max_lpool_bal);
+        return ();
+}
+
+@external
+func set_max_opt_size_perc{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
```
Comment (MarekHauzr) [inline contracts/state.cairo:395]: `set_max_option_size_percent_of_voladjspd` the getters and setters should be `get_exact_storage_var_name` and `set_exact_storage_var_name`

File: contracts/state.cairo [hunk 18]
```diff
@@ -360,6 +368,51 @@ func get_pool_volatility_auto{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, ra
...
+        assert_nn(max_opt_size_as_perc_of_vol_adjspd);
+    }
+
+    max_option_size_percent_of_voladjspd.write(max_opt_size_as_perc_of_vol_adjspd);
+    
+    return ();
+}
+
+@view
+func get_max_opt_size_perc{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() -> (res: Int){
```
Comment (MarekHauzr) [inline contracts/state.cairo:411]: ditto
